collections-arrays/HowToEfficientlyRandomizeAnArray/HowToEfficientlyRandomizeAnArray/ArrayBenchmarks.cs

```diff
@@ -0,0 +1,43 @@
+﻿using BenchmarkDotNet.Attributes;
+using BenchmarkDotNet.Mathematics;
+
+namespace HowToEfficientlyRandomizeAnArray
+{
+    [MemoryDiagnoser(true)]
+    public class ArrayBenchmarks
+    {
+        private int[] _guidArray;
+        private int[] _randomArray;
+        private int[] _fisherYatesArray;
+        private const int ARRAY_SIZE = 100000;
+        private ArrayFunctions _arrayFunctions;
+
+
+        [GlobalSetup]
+        public void Setup()
+        {
+            _arrayFunctions = new ArrayFunctions();
+            _guidArray = _arrayFunctions.GerOrderedArray(ARRAY_SIZE);
+            _randomArray = _arrayFunctions.GerOrderedArray(ARRAY_SIZE);
+            _fisherYatesArray = _arrayFunctions.GerOrderedArray(ARRAY_SIZE);
+        }
+
+        [Benchmark]
```

Comment:
Note that our benchmarks here are not an apples to oranges comparison and are showing a skewed result both in performance and memory allocation. With the Guid and Random methods we are using LINQ and returning a copy of the original array. In the FisherYates we are modifying the array in place. To make the comparisons fair and to make our helper library fair and consistent, we should modify the FisherYates method to first make a copy of the array and then operate on the copy and return the copy.
We could add a second FisherYates method also that we call InPlace, thus making it clear that it will operate on the array directly.
Something else that might be helpful to point out in the article would be some different use cases, as in some cases using the LINQ method and operating on the IEnumerable may be preferable so that we don't 1. Mutate our original array, and 2. We don't have to pay the memory cost of copying the array. Maybe we are just trying to grab the first 10 randomly ordered items from a collection for a test, or for a card game implementation of some kind (although then we probably are looking at Cryptographic random generators, but I digress).

Reply:
That is exactly what I would expect in terms of performance, the FisherYates algorithm is very efficient for shuffling, what I was focused more on is in the memory usage. You can still point out the efficiency of the algorithm, but in the article it would be good to point out that it is operating in place and thus mutating the array that is passed in. That could be made clearer in the message signature by making the method void and making it a ref parameter, which would force the caller to add a ref and thus think about the fact that this is modifying the array. That is probably outside the scope of the article, but for a benchmark, it is always good to, as much as possible, make the comparisons equal. If one method returns a copy, then the other should return a copy, otherwise the benchmark is skewed/biased in one direction as there is always going to be more memory usage for the copy and there will of course be the time necessary to do the copy.

Reply:
Great, I've implemented the method that copies the array as well. Thanks again!